### ISSUE TITLE
kube-dns --federations flag needs to be user modifiable (integration between federation and the new addon manager)

### ISSUE THREAD

**Issue by bowei** [kubernetes/kubernetes#36194]
## Overview

kube-dns currently receives a list of federation names from a command line flag `--federations`. Setup of a federated set of clusters requires the user to be able to modify the flag (see http://kubernetes.io/docs/admin/federation/#updating-kubedns).

In v1.5, the addon manager was changed (see 0e4e54a6b658fef90194ebc9e910083eccf8f29b) to periodically reconcile the configuration of the addon with a template yaml. The old behavior was to not modify the configuration if the service was present.

The --federation flag will be overwritten by the addon manager unless the template in the add-on manager itself is modified. Users of GKE cannot modify the addon manager configuration and hence will not be able to fix this behavior. User modifications in the past were preserved so this was not surfaced as an issue.

## Impact

Customers using GKE and federation will not be able to see DNS entries for the federated service instances.

## Proposed fix

- Add flag `--config-map` to kube-dns.
- Have kube-dns receive the value of --federations flag from a config map if the config map is present. The config map will have a name `kube-dns` and be in namespace `kube-system`. If the config map is not present, we default to the existing behavior which is to take the value from the command line flag.
- Have kube-dns poll for the config map change.
- Federation will either update the documentation regarding the manual updates (i.e. creation of the config map) or through the new `kubefed` CLI tool (see 2f84a7c8d2f46d52780c585024ff1038bac0a246).

**Comment by thockin** [comment]
ACK.  We've wanted a DNS config-map for a while, this is a first step.  Getting all the ComponentConfig schema plumbed through seems challenging on this timeframe - I assume the plan is to start with something compatible but incomplete?

**Comment by nikhiljindal** [comment]
cc @kubernetes/sig-cluster-federation

**Comment by bowei** [comment]
From discussion offline: with @MrHohn @kubernetes/sig-cluster-federation

There seems to be an easier solution to fix this for the 1.5 time frame:

Instead of taking the value of the command line flag directly, we can change the command line flags for kube-dns to be command:

````
command: [ "/bin/sh", "-c", "/kube-dns --federations \"$(KUBE_DNS_FEDERATIONS)\" --v=0 ..."]
````

On non-federation deployments, this will empty string, which is the default unset value.

For federation deployments, instead of editing the command line flags directly in the deployment, we have the user edit the environment variable KUBE_DNS_FEDERATIONS:
````
env:
- name: KUBE_DNS_FEDERATIONS
  value: <federations>
````
Then the value will be picked up and injected into the pod.

The reason this works is that the new addon manager only /overwrites values present in the template/. Thus, the environment variable KUBE_DNS_FEDERATION will remain editable by the user and persist.

Note: this is not the recommended longer term solution, but is being proposed to avoid a large code change in the 1.5 release.

**Comment by nikhiljindal** [comment]
This is great! Doesnt require any code change to kube-dns. Just a small change to the deployment spec.

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @bowei @nikhiljindal

**Comment by dims** [comment]
@nikhiljindal all issues must be labeled either release blocker or non release blocking by end of day 18 November 2016 PST. (please add the appropriate label or please move it to 1.6) cc @kubernetes/sig-cluster-federation

**Comment by madhusudancs** [comment]
This is a release blocker. The main PR that implemented this feature/fix was merged last night (PR #36775). The PR also has good test coverage. I will let @bowei to close this.

**Comment by nikhiljindal** [comment]
Sent PR https://github.com/kubernetes/kubernetes.github.io/pull/1788 to update documentation for this
